### PR TITLE
[move-stdlib/diem-framework] Support multiple error category modules

### PR DESCRIPTION
**Pull request by tzakian**
This PR adds support in error map generation for multiple Move modules to declare the error categories.

This PR then splits the `Errors` module two: one for the Move stdlib, and the other for the Diem Framework containing DF-specific error categories which will help with the split of the Move stdlib and Diem Framework.

### CODE REVIEW

File: language/diem-framework/modules/DiemErrors.move [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+///
+///  2. The *error reason* which is encoded in the remaining 56 bits of the code. The reason is a unique
+///     number relative to the module which raised the error and can be used to obtain more information about
+///     the error at hand. It is mostly used for diagnosis purposes. Error reasons may change over time as the
+///     framework evolves.
+///
+/// >TODO: determine what kind of stability guarantees we give about reasons/associated module.
+module DiemErrors {
+    /// A function to create an error from from a category and a reason.
+    fun make(category: u8, reason: u64): u64 {
```
Comment (sblackshear) [inline language/diem-framework/modules/DiemErrors.move:19]: Should this reuse `Errors::make`, or is there some value in keeping these totally independent?

**Comment by bob-wilson** [comment]
@tzakian Can you comment about the compatibility of this? Is this really a drop-in replacement for the current error module? I don't have a particular concern but it's touching a bunch of files and it made me wonder.

File: language/move-stdlib/modules/Errors.move [hunk 2]
```diff
@@ -1,6 +1,6 @@
 address 0x1 {
 
-/// Module defining error codes used in Move aborts throughout the framework.
+/// Module defining structured error codes in Move aborts.
 ///
 /// A `u64` error code is constructed from two values:
 ///
```
Comment (vgao1996) [inline language/move-stdlib/modules/Errors.move:6]: > Error categories are declared in this module and are globally unique across the Diem framework

Could you remove the reference to Diem Framework in a comment below?

**Review by vgao1996 — COMMENTED** [review]
Overall it looks good to me.

I think this should also be moved into the Diem Framework
```
/// A limit on an amount, e.g. a currency, is exceeded. Example: withdrawal of money after account limits window
    /// is exhausted.
    const LIMIT_EXCEEDED: u8 = 8;
```

Also this one? (IMO capability is a pattern used by the Diem Framework, but I could be wrong)
```
/// The signer of a transaction does not have a required capability.
    const REQUIRES_CAPABILITY: u8 = 4;
```

This one's probably fine, but maybe we should not reference "signing key" here?
```
/// An argument provided to an operation is invalid. Example: a signing key has the wrong format.
    const INVALID_ARGUMENT: u8 = 7;
```

File: language/diem-framework/modules/DiemErrors.move (hunk 1 above)
Comment (tzakian) [reply language/diem-framework/modules/DiemErrors.move:19]: Good question. I think making the base `Errors::make` public and using that makes sense here. I'll update.

**Comment by tzakian** [comment]
> @tzakian Can you comment about the compatibility of this? Is this really a drop-in replacement for the current error module? I don't have a particular concern but it's touching a bunch of files and it made me wonder

Yes, this is a drop-in replacement for the current error module. The reason a number of files changed is because wherever we had `Errors::requires_roles(err)` we now need to have `DiemErrors::requires_role(err)`. But this is an internal/code-only change and should not effect end-users.

**Comment by tzakian** [comment]
> Overall it looks good to me.
> 
> I think this should also be moved into the Diem Framework
> 
> ```
> /// A limit on an amount, e.g. a currency, is exceeded. Example: withdrawal of money after account limits window
>     /// is exhausted.
>     const LIMIT_EXCEEDED: u8 = 8;
> ```

I can change the example, but I actually think this should stay in the Move `Errors` module as this is not a Diem-Framework-specific error condition, but generally applicable to Move programs. E.g., when performing arithmetic we frequently do something like:
```
// a and b are u64s
assert((a as u128)  + (b as u128) < U64_MAX,  Errors::limit_exceeded(...));
let x = a + b;
```

> Also this one? (IMO capability is a pattern used by the Diem Framework, but I could be wrong)
> 
> ```
> /// The signer of a transaction does not have a required capability.
>     const REQUIRES_CAPABILITY: u8 = 4;
> ```

I can remove it, the issue is in the `Vault` code which uses this, but it makes sense to not be there. I'll work on removing this.

> This one's probably fine, but maybe we should not reference "signing key" here?
> 
> ```
> /// An argument provided to an operation is invalid. Example: a signing key has the wrong format.
>     const INVALID_ARGUMENT: u8 = 7;
> ```

I'll update the comment :)

**Comment by wrwg** [comment]
While first seeing this, I thought this is a fine pragmatic approach to the problem of error codes, but I'm not that sure any longer.

I tend to think that our current error code names like ROLE_REQUIRED are unnecessary specific. This should be instead ACCESS_DENIED or similar. The applications consuming these error codes don't care how we currently implement access control.

I wonder whether we can just generalize the error naming and then would not even need a `DiemErrors` module. A canonical error space is very beneficial, if it is possible.

**Comment by tzakian** [comment]
> While first seeing this, I thought this is a fine pragmatic approach to the problem of error codes, but I'm not that sure any longer.
> 
> I tend to think that our current error code names like ROLE_REQUIRED are unnecessary specific. This should be instead ACCESS_DENIED or similar. The applications consuming these error codes don't care how we currently implement access control.

Good point. I think doing this would make a lot of sense (`REQUIRES_ROLE` => `ACCESS_DENIED`, `REQUIRES_CAPABILITY` => `NO_PERMISSION`). I also agree on providing a canonical error space for these if possible.

However, one thing that I worry is that in the responses from json-rpc that contain the error explanations, the category name is provided. So I'm worried that changing the names of these variables would constitute a breaking change to clients.

**Comment by tzakian** [comment]
@wrwg and @vgao I put up an alternate PR for this in #8587 that keeps a single errors category module, preserves backwards compatibility, and generalizes the category names (would be interested on any thoughts on if there is a better way to handle `requires_capability` in that PR).

**Pull request closed without merging**